Performance.md

```diff
@@ -0,0 +1,290 @@
+              ------------
+              |          |
+              |  Shared  |
+              ^----------^
+             /            \
+            /              \
+------------                ------------
+|          |                |          |
+|  Client  |                |  Server  |
+-----^------                ------^-----
+```
+
+Tests can also be broken into their own project.
+
+```
+              ------------
+              |          |
+              |  Shared  |
+              ^-----^----^
+             /      |     \
+            /       |      \
+------------  ------------  ------------
+|          |  |  Shared  |  |          |
+|  Client  |  |  Tests   |  |  Server  |
+-----^------  ------------  ------^-----
+     |                            |
+     |                            |
+------------                ------------
+|  Client  |                |  Server  |
+|  Tests   |                |  Tests   |
+------------                ------------
+```
+
+You can read up more about project references here. <!-- TODO -->
+
+# Configuring `tsconfig.json`
+
+Within a `tsconfig.json`, there are two ways to specify files in a project:
+
+* the `files` list
+* the `include` and `exclude` lists
+
+The primary difference between the two is that `files` expects a list of file paths to source files, and `include/`exclude` use globbing patterns to match against files.
+
+While specifying `files` will allow TypeScript to quickly load up files up directly, it can be cumbersome if you have many files in your project without just a few top-level entry-points.
+Additionally, it's easy to forget to add new files to your `tsconfig.json`, which means that you might end up with strange editor behavior where those new files are incorrectly analyzed.
+All this can be cumbersome.
+
+`include`/`exclude` help avoid needing to specify these files, but at a cost: files must be discovered by walking through included directories.
+When running through a *lot* of folders, this can slow compilations down.
+Additionally, sometimes a compilation will include lots of unnecessary `.d.ts` files and test files, which can increase compilation time and memory overhead.
+Finally, while `exclude` has some reasonable defaults, certain configurations like mono-repos mean that a "heavy" folders like `node_modules` can still end up being included.
+
+For best practices, we recommend the following:
+
+* Specify only source folders in your project.
+* Don't mix source files from other projects in the same folder.
+* If keeping tests in the same folder as other source files, give them a distinct name so they can easily be excluded.
+* Avoid large build artifacts and dependency folders like `node_modules` in source directories
+
+Here is a reasonable `tsconfig.json` that demonstrates this in action.
+
+```json5
+{
+    "compilerOptions": {
+        // ...
+    },
+    "include": ["src"],
+    "exclude": ["**/node_modules", "**/.*/"],
+}
+```
+
+## Incremental Project Emit
```

Comment:
I don't know where it is right to mention but there are these two points we should mention

When people notice incremental or --watch taking longer.. It could be because of .d.ts takes longer in those scenarios.. The way to diagnose that is to run `tsc --emitDeclarationOnly` and see the emitTime.. is it excessive? that could be indication that the .d.ts emit is not recent.. And just now @weswigham replied on a thread with https://github.com/microsoft/TypeScript/issues/34119#issuecomment-545542685 and that could also be part of the guide....

Another issue could be a bug (or not but to turn incremental on and run `tsc --extendedDiagnostics` to see if checkTime is exceptionally heavy when there are no or small changes.. People reporting those time can clue us in we have a issue - being checking more files than necessary or it could be the issue with the project setup...